src/Perlang.Parser/NumberParser.cs

```diff
@@ -0,0 +1,141 @@
+                    BigInteger.Parse('0' + numberCharacters, numericToken.NumberStyles),
+
+                _ =>
+                    throw new InvalidOperationException($"Base {(int)numericToken.NumberBase} not supported")
+            };
+
+            if (value <= Int32.MaxValue)
+            {
+                return (int)value;
+            }
+            else if (value <= UInt32.MaxValue)
+            {
+                return (uint)value;
+            }
+            else if (value <= Int64.MaxValue)
+            {
+                return (long)value;
+            }
+            else if (value <= UInt64.MaxValue)
+            {
+                return (ulong)value;
+            }
+            else // Anything else remains a BigInteger
+            {
+                return value;
+            }
+        }
+    }
+
+    public static object MakeNegative(object value)
+    {
+        if (value is double doubleValue)
+        {
+            return -doubleValue;
+        }
+        else if (value is int doubleInt)
+        {
+            return -doubleInt;
+        }
+        else if (value is uint doubleUint)
+        {
+            long negativeValue = -doubleUint;
+
+            // This is a special hack to ensure that the value -2147483648 gets returned as an `int` and not a `long`.
+            // Some details available in #302, summarized here in brief:
+            //
+            // The value 2147483648 is too large for an `int` => gets parsed into a `ulong` where it will fit. Once it
+            // has been made negative, the value -2147483648 is again small enough to fit in an `int` => the code below
+            // will narrow it down to comply with the "smallest type possible" design principle.
+            //
+            // Rationale: Two's complement: https://en.wikipedia.org/wiki/Two%27s_complement
+            if (negativeValue >= Int32.MinValue)
+            {
+                return (int)negativeValue;
+            }
+            else
+            {
+                return negativeValue;
+            }
+        }
+        else if (value is long longValue)
+        {
+            return -longValue;
+        }
+        else if (value is ulong ulongValue)
+        {
+            // Again, this needs to be handled specially to ensure that numbers that fit in a `long` doesn't use
+            // BigInteger unnecessarily.
+            BigInteger negativeValue = -new BigInteger(ulongValue);
+
+            if (negativeValue >= Int64.MinValue)
+            {
+                return (long)negativeValue;
+            }
+            else
+            {
+                // All negative numbers that are too big to fit in any of the smaller signed integer types will go
+                // through this code path.
+                return negativeValue;
+            }
+        }
+        else
+        {
+            throw new ArgumentException($"Type {value.GetType().ToTypeKeyword()} not supported");
+        }
+    }
+}
```

Comment:
Extracting all of this into a separate class has the nice side effect of making it more easy to test this in isolation in a unit test, if needed. Right now, there's no pressing need; we have integration tests which ensures this code gets exercised thoroughly. When we add support for new integer types, it could make sense to add unit tests if it makes sense. In general, I feel that integration tests have the greatest value in this kind of project anyway, so this is just a vague idea/suggestion.